horovod/runner/http/http_server.py

```diff
@@ -122,13 +122,29 @@ def do_DELETE(self):
 
         _, scope, key = paths
 
-        with self.server.finished_list_lock:
-            self.server.finished_list[scope].append(key)
-            if scope in self.server.scope_size:
-                # scope is not an "ad-hoc" scope added by a client
+        if scope in self.server.scope_size:
+            # scope defined by host_alloc_plan at server initialization
+            with self.server.finished_list_lock:
+                self.server.finished_list[scope].append(key)
                 if self.server.scope_size[scope] == len(self.server.finished_list[scope]):
                     with self.server.cache_lock:
                         self.server.cache.get(scope, {}).clear()
+                        self.server.finished_list.clear()
```

Comment:
`finished_list` was never cleared before this change. 

The list is supposed to track all clients that requested their key to be deleted (which happens at the end of the Rendezvous process). Once the list is full, the cache for this scope is cleared. This worked fine for the first Rendezvous process, but if Horovod was shut down and reinitialized, the list would grow too long and the cache would never be cleared anymore.

I am not sure if this detail might be related to the occasional Gloo timeouts I've seen in `test_horovod_reinit()`. 

Also I haven't understood yet why a separate finished_list is tracked here and key-value pairs aren't simply popped off the cache right away (it seems that there may be situations where a GET after a supposedly successful DELETE for the same key would still return a value).

Reply:
> Also I haven't understood yet why a separate finished_list is tracked here and key-value pairs aren't simply popped off the cache right away (it seems that there may be situations where a GET after a supposedly successful DELETE for the same key would still return a value).

OK, I've figured out this bit. `Rendezvous()` in gloo_context.cc calls `Finalize()` on the HTTPStore potentially before all clients have connected. So a GET after a DELETE is actually supposed to return a value, else there will be problems:
```
[1]<stdout>:[2021-06-22 16:19:12.630567: T /mnt/data/max/horovod/horovod/common/gloo/http_store.cc:132] Send DELETE request to http://127.0.0.1:43444/local_localhost_process_set_hash_1e9d548172549547/1
[0]<stdout>:[2021-06-22 16:19:12.630843: T /mnt/data/max/horovod/horovod/common/gloo/http_store.cc:103] Send GET request to http://127.0.0.1:43444/local_localhost_process_set_hash_1e9d548172549547/1
Scope local_localhost_process_set_hash_1e9d548172549547: Deleted key 1
scope local_localhost_process_set_hash_1e9d548172549547: get 1 -> None
[0]<stderr>:terminate called after throwing an instance of 'gloo::EnforceNotMet'
```
I need to be more careful to defer the actual deletion.

Reply:
Rather than introduce more complexity to the `RendezvousServer` in Python, I just put a Gloo barrier into `Rendezvous()`, which generally feels like a sane thing to have there to me. 

This seems to have cleared up the weird issues with the tests!

Reply:
@tgaddair, this synchronization barrier that I stuck into `Rendezvous()` and clearing the `finished_list` in the rendezvous server are the biggest changes to the general Gloo code coming with this PR.